libsolidity/codegen/CompilerUtils.cpp

```diff
@@ -923,24 +929,76 @@ void CompilerUtils::convertType(
 				// stack: <mem start> <source ref> (variably sized) <length> <mem data pos>
 				if (targetType.baseType()->isValueType())
 				{
-					solAssert(typeOnStack.baseType()->isValueType(), "");
 					copyToStackTop(2 + stackSize, stackSize);
+					// stack: <mem start> <source ref> (variably sized) <length> <mem data pos> <source ref> (variably sized)
 					ArrayUtils(m_context).copyArrayToMemory(typeOnStack);
 				}
 				else
 				{
+					// Since the base type is not a value type, we need recursively convert each array element
 					m_context << u256(0) << Instruction::SWAP1;
 					// stack: <mem start> <source ref> (variably sized) <length> <counter> <mem data pos>
 					auto repeat = m_context.newTag();
 					m_context << repeat;
 					m_context << Instruction::DUP3 << Instruction::DUP3;
+					// stack: <mem start> <source ref> (variably sized) <length> <counter> <mem data pos> <length> <counter>
 					m_context << Instruction::LT << Instruction::ISZERO;
 					auto loopEnd = m_context.appendConditionalJump();
+					// stack: <mem start> <source ref> (variably sized) <length> <counter> <mem data pos>
 					copyToStackTop(3 + stackSize, stackSize);
+					// stack: <mem start> <source ref> (variably sized) <length> <counter> <mem data pos> <source ref> (variably sized)
 					copyToStackTop(2 + stackSize, 1);
-					ArrayUtils(m_context).accessIndex(typeOnStack, false);
-					if (typeOnStack.location() == DataLocation::Storage)
+					// stack: <mem start> <source ref> (variably sized) <length> <counter> <mem data pos> <source ref> (variably sized) <counter>
+					switch (typeOnStack.location())
+					{
+					case DataLocation::Storage:
+						ArrayUtils(m_context).accessIndex(typeOnStack, false, false);
 						StorageItem(m_context, *typeOnStack.baseType()).retrieveValue(SourceLocation(), true);
+						break;
+					case DataLocation::CallData:
```

Comment:
Copied from ExpressionCompiler's indexAccess implementation

Reply:
Can you extract it to a public memory function if ``CompilerUtils`` then and call it from ``ExpressionCompiler::visit(IndexAccess const&)``?

Reply:
Ok